### PR TITLE
Optimize X86 minmaxEvaluator

### PR DESCRIPTION
**Pull request by 0dvictor**
Optimized X86 minmaxEvaluator using CMOVcc instructions.

Signed-off-by: Victor Ding <dvictor@ca.ibm.com>

### CODE REVIEW

**Comment by vijaysun-omr** [comment]
I think reviews are needed from @andrewcraik and @0xdaryl for this one.

**Comment by andrewcraik** [comment]
LGTM from me

File: compiler/x/codegen/ControlFlowEvaluator.cpp [hunk 1]
```diff
@@ -524,103 +524,98 @@ TR::Register *OMR::X86::TreeEvaluator::tableEvaluator(TR::Node *node, TR::CodeGe
...
-   node->setRegister(trgReg);
+   return result;
+   }
+TR::Register *OMR::X86::TreeEvaluator::minmaxEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+   {
+   if (TR::Compiler->target.is32Bit() && 
+       (node->getOpCodeValue() == TR::lmin || node->getOpCodeValue() == TR::lmax))
+      {
+      return minmaxEvaluator64BitPrimitiveOnIA32(node, cg);
+      }
```
Comment (0xdaryl) [inline compiler/x/codegen/ControlFlowEvaluator.cpp]: I think the cleaner way to do this is to modify the i386 TreeEvaluatorTable (`x/i386/codegen/TreeEvaluatorTable.hpp`) for the `TR::lmin` and `TR::lmax` opcodes to point directly to the `minmaxEvaluator64BitPrimitiveOnIA32` static function.  This redirection wouldn't be necessary here then.

File: compiler/x/codegen/ControlFlowEvaluator.cpp [hunk 2]
```diff
@@ -524,103 +524,98 @@ TR::Register *OMR::X86::TreeEvaluator::tableEvaluator(TR::Node *node, TR::CodeGe
...
+         SETcc = SETL1Reg;
+         break;
+      case TR::lmax:
+         SETcc = SETG1Reg;
+         break;
+      default:
+         TR_ASSERT(false, "INCORRECT IL OPCODE.");
+         break;
+      }
+   auto oprand0 = cg->evaluate(node->getChild(0));
```
Comment (0xdaryl) [inline compiler/x/codegen/ControlFlowEvaluator.cpp]: I'm not sure if its intentional to abbreviate `operand` by one letter or if its just a spelling mistake, but I think it reads better as `operand`.  Please change throughout this function.

File: compiler/x/codegen/ControlFlowEvaluator.cpp [hunk 3]
```diff
@@ -524,103 +524,98 @@ TR::Register *OMR::X86::TreeEvaluator::tableEvaluator(TR::Node *node, TR::CodeGe
...
+   for (int i = 0; i < node->getNumChildren(); i++)
       {
-      generateRegRegInstruction(cmp_op, node, reg, trgReg, cg);
-      generateLabelInstruction(isMin ? JG4 : JL4, node, doneLabel, false, cg);
-      generateRegRegInstruction(move_op, node, trgReg, reg, cg);
-
-      dep = new (cg->trHeapMemory()) TR::RegisterDependencyConditions(0, 2, cg->trMemory());
-      dep->addPostCondition(trgReg, TR::RealRegister::NoReg, cg);
-      dep->addPostCondition(reg, TR::RealRegister::NoReg, cg);
+      cg->decReferenceCount(node->getChild(i));
```
Comment (0xdaryl) [inline compiler/x/codegen/ControlFlowEvaluator.cpp]: Is a loop necessary?  Can `min` or `max` ever have anything other than 2 children?  I'd say just do two `decReferenceCount`s to make the code easier to read.

File: compiler/x/codegen/ControlFlowEvaluator.cpp [hunk 4]
```diff
@@ -524,103 +524,98 @@ TR::Register *OMR::X86::TreeEvaluator::tableEvaluator(TR::Node *node, TR::CodeGe
...
+         CMP = CMP8RegReg;
+         MOV = MOV8RegReg;
+         CMOV = CMOVL8RegReg;
+         break;
+      default:
+         TR_ASSERT(false, "INCORRECT IL OPCODE.");
+         break;
+      }
+   
+   auto oprand0 = cg->evaluate(node->getChild(0));
```
Comment (0xdaryl) [inline compiler/x/codegen/ControlFlowEvaluator.cpp]: I'm not sure if its intentional to abbreviate `operand` by one letter or if its just a spelling mistake, but I think it reads better as `operand`.  Please change throughout this function.

File: compiler/x/codegen/ControlFlowEvaluator.cpp [hunk 5]
```diff
@@ -524,103 +524,98 @@ TR::Register *OMR::X86::TreeEvaluator::tableEvaluator(TR::Node *node, TR::CodeGe
...
+   auto result = cg->allocateRegister();
+   generateRegRegInstruction(CMP,  node, oprand0, oprand1, cg);
+   generateRegRegInstruction(MOV,  node, result,  oprand0, cg);
+   generateRegRegInstruction(CMOV, node, result,  oprand1, cg);
+
+   node->setRegister(result);
    for (int i = 0; i < node->getNumChildren(); i++)
       {
       cg->decReferenceCount(node->getChild(i));
       }
```
Comment (0xdaryl) [inline compiler/x/codegen/ControlFlowEvaluator.cpp]: Is a loop necessary?  Can `min` or `max` ever have anything other than 2 children?  I'd say just do two `decReferenceCount`s to make the code easier to read.

**Comment by 0xdaryl** [comment]
This PR touches files built on x86 only.  A genie build is not required.

File: compiler/x/codegen/ControlFlowEvaluator.cpp (hunk 2 above)
Comment (0dvictor) [reply compiler/x/codegen/ControlFlowEvaluator.cpp]: It was a typo. Now it has been fixed.

File: compiler/x/codegen/ControlFlowEvaluator.cpp (hunk 3 above)
Comment (0dvictor) [reply compiler/x/codegen/ControlFlowEvaluator.cpp]: Sure, I've updated this part to two `decReferenceCount`

File: compiler/x/codegen/ControlFlowEvaluator.cpp (hunk 1 above)
Comment (0dvictor) [reply compiler/x/codegen/ControlFlowEvaluator.cpp]: Good suggestion, updated.

File: compiler/x/codegen/ControlFlowEvaluator.cpp (hunk 4 above)
Comment (0dvictor) [reply compiler/x/codegen/ControlFlowEvaluator.cpp]: It was a typo, not it has been fixed.

File: compiler/x/codegen/ControlFlowEvaluator.cpp (hunk 5 above)
Comment (0dvictor) [reply compiler/x/codegen/ControlFlowEvaluator.cpp]: Sure, the ILOp should only have 2 children. I've updated this part to two `decReferenceCount`

**Comment by vijaysun-omr** [comment]
Reviews are complete. I am merging.

**Pull request merged**